traffic_ops/traffic_ops_golang/deliveryservice/letsencryptcert.go

```diff
@@ -0,0 +1,314 @@
+package deliveryservice
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/config"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/riaksvc"
+	"github.com/go-acme/lego/certcrypto"
+	"github.com/go-acme/lego/certificate"
+	"github.com/go-acme/lego/challenge"
+	"github.com/go-acme/lego/challenge/dns01"
+	"github.com/go-acme/lego/lego"
+	"github.com/go-acme/lego/registration"
+	"github.com/jmoiron/sqlx"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type MyUser struct {
+	Email        string
+	Registration *registration.Resource
+	key          crypto.PrivateKey
+}
+
+func GetLetsEncryptTimeout() time.Duration {
+	return time.Minute * 10
+}
+
+func (u *MyUser) GetEmail() string {
+	return u.Email
+}
+
+func (u MyUser) GetRegistration() *registration.Resource {
+	return u.Registration
+}
+
+func (u *MyUser) GetPrivateKey() crypto.PrivateKey {
+	return u.key
+}
+
+type DNSProviderTrafficRouter struct {
+	db *sqlx.DB
+}
+
+func NewDNSProviderTrafficRouter() (*DNSProviderTrafficRouter, error) {
+	return &DNSProviderTrafficRouter{}, nil
```

Comment:
done. removed the error. it cant be passed into Present and Cleanup because those use the lego Provider interface which only allows domain, token, and keyAuth strings as parameters